oozie-to-airflow/mappers/pig_mapper.py

```diff
@@ -27,7 +27,7 @@
 from utils.template_utils import render_template
 
 
-class PigMapper(ActionMapper, PrepareMixin, ArchiveMixin, FileMixin):
+class PigMapper(PrepareMixin, ArchiveMixin, FileMixin, ActionMapper):
```

Comment:
Are you sure that this change does not breaks anything?

Reply:
I made it to actually fix something. When the `ActionMapper` was first, the `on_parse_node()` was invoked on `BaseMapper`, which does nothing.
This change allows the `on_parse_node()` of `PrepareMixin` and `ArchiveMixin` to work.

Reply:
Or maybe the `ActionMapper` should implement:

```
    def on_parse_node(self):
        super().on_parse_node()
```
?

Reply:
I have another solution.
Please look at: https://termbin.com/c3lw
This forces the correct MRO of PigMapper.
To add my commit, please run a command on this branch:
```
curl https://termbin.com/c3lw | git am
```

Reply:
Your MRO:
```
(
<class 'mappers.pig_mapper.PigMapper'>, 
<class 'mappers.prepare_mixin.PrepareMixin'>, 
<class 'mappers.file_archive_mixins.ArchiveMixin'>, 
<class 'mappers.file_archive_mixins.FileMixin'>, 
<class 'mappers.action_mapper.ActionMapper'>, 
<class 'mappers.base_mapper.BaseMapper'>, 
<class 'object'>
)
```

Reply:
Thanks, good suggestion! I've applied your fix (I forgot about the fixup, sorry - I squashed it into 1 commit).

Reply:
This solution is problematic because when FileMixin and ArchiveMixin inherit from BaseMapper, they have to implement quite a few unnecessary methods. This is a leaking abstraction.
Now I think that ideally FileMixin and ArchiveMixin should be changed to something like FileParser and ArchiveParser and used when needed using composition. This would make the design cleaner.

However, due to time constraints right now, I've decided to go back to my original "workaround" and we'll refactor this in the future.

Reply:
I've created a relevant issue in Github so that we don't forget about this: https://github.com/GoogleCloudPlatform/cloud-composer/projects/1#card-20376939

Reply:
@sprzedwojski Where FileMixin and ArchiveMixin need implementation of methods? It seems to me that only in tests. You can not create an object of this class, but you should not do it.  Corresponding tests should be in classes that use this mixin.  Mixin is not a real class, only a structure that allows you to write less code.

For example:
You can write mixin, which manages permissions (https://github.com/django/django/blob/master/django/contrib/auth/mixins.py#L47-L52), but You should write tests if the class uses this mixin.

Reply:
The problem is similar to testing abstract classes in Java. Mixin, which inherits from BaseMapper, is an abstract class. 
This code is incorrect in this situations.
https://github.com/GoogleCloudPlatform/cloud-composer/blob/master/oozie-to-airflow/tests/mappers/test_archive_mixin.py#L50

Reply:
After looking at the discussion  and seeing the MRO in action. I think it is a bit too "magical" solution with the mixins and calling super(). I think that falls into "explicit is better than implicit" zen as it's not obvious at all how it is working with MRO and multiple Mixins. 

I think I favour @sprzedwojski proposal of using composition rather than inheritance. It's far more explicit and straightforward IMHO and has much less magic involved.

Reply:
@mik-laj If you run our pre-commit hook it complains that not all methods of BaseMapper are implemented in File and ArchiveMixins if they inherit from BaseMapper.
I'd leave it as is for now and move on. We have an issue for it in Github so we won't forget about it. Let's refactor it in May.

Reply:
Refactored to composition